packages/vm/src/evm/opcodes/util.ts

```diff
@@ -334,9 +335,38 @@ export const eof1CodeAnalysis = (container: Buffer) => {
     }
     if (container.length !== computedContainerSize) {
       // Scanned code does not match length of contract byte code
-
       return
     }
     return sectionSizes
   }
 }
+
+export const eof1ValidOpcodes = (code: Buffer) => {
+  // EIP-3670 - validate all opcodes
+  const opcodes = new Set(handlers.keys())
+  opcodes.add(0xfe) // Add INVALID opcode to set
+
+  let x = 0
+  while (x < code.length) {
+    const opcode = code[x]
+    x++
+    if (!opcodes.has(opcode)) {
+      // No invalid/undefined opcodes
+      return false
+    }
+    if (opcode >= 0x60 && opcode <= 0x7f) {
+      // Skip data block following push
+      x += opcode - 0x5f
+      if (x > code.length - 1) {
+        // Push blocks mmust not exceed end of code section
+        return false
+      }
+    }
+  }
+  const terminatingOpcodes = new Set([0x00, 0xd3, 0xfd, 0xfe, 0xff])
```

Comment:
I guess a dedicated list why each of the opcodes here is regarded as "terminating" would be good as well.

Reply:
I expanded the comment to clearly identify the terminating opcodes (STOP, RETURN, REVERT, INVALID, and SELFDESTRUCT) as identified in the EIP.  Not sure there's much more to the logic of why each of these specifically other than these are the only opcodes that logically terminate a sequence of instructions.  Is this what you meant?

Reply:
Ah, yes, that works respectively gives enough context I guess, thanks! 😄